### PR TITLE
CloudEvents SDK V2 & Eventing Channel Updates

### PR DESCRIPTION
**Pull request by travis-minke-sap**
This PR is for Issue #7 and is expected to be the final duplicate delivery from the kyma-incubator repository ([knative-kafka](https://github.com/kyma-incubator/knative-kafka)) project.

The changes are to move the implementation to the CloudEvents SDK V2 and to align with more recent versions of the Knative eventing Channel framework.

The new files (retry*.go) are simply a refactoring of the removed `client` package - the logic therein is the same.

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
Hi @davyodom , i don't see on this pr that you're using the `binding.Message`/`binding.MessageWriter` constructs to marshal/unmarshal the message, is that something you're planning to do later?

**Comment by travis-minke-sap** [comment]
Hi slinkydeveloper, 

In the channel's `main.go` we are using the `binding.ToEvent()` to convert the `Message` into a cloud event.  The current implementation is/was based on cloud events and since there currently isn't any protocol for kafka_confluent this was the easiest transition.  Similarly, in the dispatcher's `retry.go` file we are using the `binding.ToMessage()` to convert our cloud event into a `Message`.  

Not sure if this addresses your question?  Let me know if there's a better approach.

**Comment by slinkydeveloper** [comment]
> In the channel's main.go we are using the binding.ToEvent() to convert the Message into a cloud event. The current implementation is/was based on cloud events and since there currently isn't any protocol for kafka_confluent this was the easiest transition. Similarly, in the dispatcher's retry.go file we are using the binding.ToMessage() to convert our cloud event into a Message.

Ok maybe i overlooked this PR :smile: BTW Why are you going to event and back? Do you really need to materialize the event? Can you just use the Message?

**Comment by travis-minke-sap** [comment]
TLDR - We already had the manual logic to go from a Confluent Message to a CloudEvent `Event` and didn't want to re-write that until we know better which kafka library we'll settle on.

Yeah - you're right it's inefficient and there are notes in the code to that effect.  This is only because the current implementation is based on Cloud Events and we're now tying into the knative messaging which is based on the sdk v2 `Messages`.  This was just the least-effort mechanism to convert between the two given the current implementation.

In the longer term one of the following will likely adjust/improve this...
1 - If the aligned implementation settles on the Sarama library, then we can use the protocol `kafka_sarama` to go directly from sarama messages to cloud event messages.
2 - If we stay with confluent the we can either look at writing another protocol `kafka_confluent` to do the same - OR - we could alter our current implementation to manually go between the Confluent Message and the CloudEvent Message instead of having this CloudEvent `Event` intermediary.

**Comment by slinkydeveloper** [comment]
>  If we stay with confluent the we can either look at writing another protocol kafka_confluent to do the same

I think this is the best solution. In the end you don't need a full protocol, but just the implementation of `Message` and `WriteTo[]` (as explained here: https://cloudevents.github.io/sdk-go/protocol_implementations.html)
Also, look at the implementation of the kafka channel in eventing-contrib to check how we leverage `Message` and base code from eventing: https://github.com/knative/eventing-contrib/blob/master/kafka/channel/pkg/dispatcher/dispatcher.go#L135

**Comment by travis-minke-sap** [comment]
Thanks for the documentation link - hadn't seen that - will take a look!

I have looked extensively at the eventing-contrib kafka (we're borrowing lots of things from there and following your lead - thanks!).  The difference is simply the fact that we're on Confluent and didn't have the protocol handy.

**Comment by slinkydeveloper** [comment]
If you want, i can contribute to that :wink:

**Comment by slinkydeveloper** [comment]
But i suppose this depends on https://github.com/knative-sandbox/eventing-kafka/issues/8

**Comment by travis-minke-sap** [comment]
Yeah - exactly - we should settle on the kafka library first - but then we'd definitely welcome any help with the protocol implementation - thanks!
